curtains/src/androidTest/java/curtains/test/utilities/TestUtils.kt

```diff
@@ -88,4 +91,27 @@ fun View.onAttachedToWindow(onAttachedToWindow: () -> Unit) {
 
     override fun onViewDetachedFromWindow(v: View) = Unit
   })
+}
+
+fun <T : Activity> launchWaitingForFocus(activityClass: KClass<T>): ActivityScenario<T> {
+  val waitForFocus = CountDownLatch(1)
+  val scenario = ActivityScenario.launch(activityClass.java)
+  scenario.onActivity { activity ->
+    if (activity.hasWindowFocus()) {
+      waitForFocus.countDown()
+    } else {
+      activity.window.onWindowFocusChangedListeners += object : OnWindowFocusChangedListener {
+        override fun onWindowFocusChanged(hasFocus: Boolean) {
+          if (hasFocus) {
+            activity.window.onWindowFocusChangedListeners -= this
+            waitForFocus.countDown()
+          } else {
+            error("expected activity to become focused")
```

Comment:
I don't know why I went with error() here vs check in the tests. Should probably use check here as well.

Note: I removed the assertThat calls because generally asserts are to be used for the verification part of a test (ie what's being tested) which is different from making sure everything is in the right state for the test to proceed. The result is similar, just different semantics which help understand what the test is about (and yes, probably some other tests need assertThat removal)

Reply:
Using the thing under test as a scaffolding around the test, huh :D

> I don't know why I went with error() here vs check in the tests

If it was in the test case, I'd consider `Assert.fail()`. If it's in helper function I'm not sure.

--- 

I had the count downs before assertions because I was afraid of having to wait 30 seconds for the failure to bubble up. The relation was inversed, the latch was just a control mechanism so the activity isn't finished prematurely and then the assertions inside the callback made sense.

How about this instead?
* I use the latch as a control mechanism, same as before, but don't assert on it.
* The listener pushes new items into a `mutableList<Boolean>()`.
* At the absolute end of each test I assert that the list contains the expected value.

Reply:
Alright, I have it cooked up, let me know when I can blow up the logging

Sneak peek:
![obrazek](https://user-images.githubusercontent.com/4589773/107575966-cff20580-6bf0-11eb-93aa-b72ccbe10564.png)

Basically, don't wait 30 seconds for the exception to bubble up just for getting unexpected `hasFocus` value. Made the same change in the helper. And `check` now makes sense where `error` was.